### PR TITLE
 Limit form errors to one at a time

### LINKED ISSUE #10063: Limit form errors to one at a time
The form Error component is showing all the possible errors based on the form control validations. That's no a pleasant example can be found for an email field, where we have the required validation as well as the valid email address pattern validation. If the user only touch the form field, we like to only show the first validation error (required), and keep the other validations for a later phase.

### PR DESCRIPTION
**Pull request by markywilson**
fixes #10063

### CODE REVIEW

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.html [hunk 1]
```diff
@@ -1,3 +1,3 @@
-<p *ngFor="let errorName of errors$ | async">
+<p *ngIf="error$ | async as errorName">
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.html]: I actually would prefer to keep the original UI implementation, and limit the number of errors in the controller. This means that it's (1) less of a breaking change and (2) an optional thing. We could even make it configurable, or at least least allow to easily override it.

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts [hunk 2]
```diff
@@ -28,16 +35,17 @@ export class FormErrorsComponent {
...
-          .filter((error) => error[1])
-          .map((error) => error[0])
-      )
-    );
+  /**
+   * The name of the control is enough
+   * @param controlName
+   */
+  @Input()
+  set controlName(controlName: string) {
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts]: I've checked once more why we can't use this component with a static controlName reference, as I believe this is quite a drop in dx. 

```html
<form-errors controlName="email"></form-errors>
```

I believe this is caused by the fact that at the time the static input is set, the form container/control  is not yet initialised. You'd therefor need to move the logic to a lifecycle hook. 

To support dynamic forms, I'd recommend to use the `OnChanges` hook, and validate whether the controlName or control are changed. With that, you can further simplify the controller: 
- use simple inputs (no getters/setters needed)
- use the public `control` only (drop private _control), and set the control only if the controlName is passed in

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts [hunk 3]
```diff
@@ -17,8 +19,13 @@ import { map, startWith } from 'rxjs/operators';
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class FormErrorsComponent {
+  _parent: FormGroupDirective;
   _control: FormControl;
-  errors$: Observable<string[]>;
+  error$: Observable<string>;
+
+  constructor(@Optional() @Host() parent: FormGroupDirective) {
+    this._parent = parent;
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts]: Why not make the argument protected? also would rename it to express the value better, i.e. 

```ts
@Optional() @Host() protected formGroup: FormGroupDirective
```

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts [hunk 4]
```diff
@@ -53,4 +51,45 @@ export class FormErrorsComponent {
   @HostBinding('class.control-touched') get touched() {
     return this.control?.touched;
   }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.control) {
+      this.control = changes.control.currentValue;
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts]: This seems not needed, right? The default behaviour of inputs work

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts [hunk 5]
```diff
@@ -16,33 +20,27 @@ import { map, startWith } from 'rxjs/operators';
...
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class FormErrorsComponent {
-  _control: FormControl;
+export class FormErrorsComponent implements OnChanges {
+  _formGroup: FormGroupDirective;
   errors$: Observable<string[]>;
 
+  constructor(@Optional() @Host() protected formGroup: FormGroupDirective) {
+    this._formGroup = formGroup;
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts]: why? We can just access `this.formgroup` in this controller, no need to store it in a local var, right?

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts [hunk 6]
```diff
@@ -16,33 +20,27 @@ import { map, startWith } from 'rxjs/operators';
   templateUrl: './form-errors.component.html',
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class FormErrorsComponent {
-  _control: FormControl;
+export class FormErrorsComponent implements OnChanges {
+  _formGroup: FormGroupDirective;
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts]: can we drop this see below? Ohterwise we should make it protected.

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts [hunk 7]
```diff
@@ -53,4 +51,45 @@ export class FormErrorsComponent {
...
+  private setErrors(): void {
+    this.errors$ = this.control?.statusChanges.pipe(
+      startWith({}),
+      map(() => this.control.errors || {}),
+      map((errors) => {
+        const filteredErrors = Object.entries(errors)
+          .filter((error) => error[1])
+          .map((error) => error[0]);
+        return this.showOnlyOne
+          ? filteredErrors[0]
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts:75]: `filteredErrors?.[0]` ?

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]
few nitpicks

File: projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts (hunk 7 above)
Comment (markywilson) [reply projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts:75]: The issue is that we need to wrap the `[0]` element in an array to be iterateable in the template.
Without the 2nd ternary operator ,`[undefined]` not an `[]` is returned.

Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/shared/components/form/form-errors/form-errors.component.ts:75]: understood. you could do `filteredErrors?.[0] ?? []` i guess.

**Pull request closed without merging**